### PR TITLE
[scrapers] python3 based artist and album scraper

### PR DESCRIPTION
**Pull request by ronie**
in PR https://github.com/xbmc/xbmc/pull/9984 spiff added support for python based metadata scrapers,
which will allow us to deprecate XML based scraping.

since i'm unsure if we're all on the same page regarding the XML deprecation, these new album & artist scrapers can be used as an alternative to their metadata.*.universal equivalents.

i'll rename them to metadata.album.universal / metadata.artists.universal if we decide to fully replace the XML scrapers.

@notspiff @olympia @DaveTBlake 

(p.s. python video scrapers are found here: PR https://github.com/xbmc/xbmc/pull/12661)

### CODE REVIEW

**Comment by DaveTBlake** [comment]
Thanks for this @ronie.   I need to find some time to get stuck into the details (e.g. extra artist data to scrape, using and storing both MB release group id and release id for albums)  before testing it against some of the more diverse music collections

>i'll rename them to metadata.album.universal / metadata.artists.universal if we decide to fully replace the XML scrapers.

I don't think a rename is the best idea. For a while these will possibly run in parallel, there is no reason not to and it makes a good test vehicle, plus the new scrapers should get their own billing. But see what others think.

**Comment by MartijnKaijser** [comment]
I also feel the rename will cause way more issues than we need. This will totally kill the smooth upgrade between v18 and v19

**Comment by olympia** [comment]
I really do hope this doesn't mean to have the whole XML scraping engine depreciated and retired. I.e. I hope XML scrapers will be supported in the future as well.

I also fully agree with renaming not being a good approach. Let's have the new scrapers default as from you feel comfortable, but also please keep XML scrapers going.

**Comment by zag2me** [comment]
This is awesome! Would be good to get a dedicated TheAudioDB scraper as well at some point to replace the current XML one. The Musicbrainz web service isn't always reliable for the artist string searches.

**Comment by ronie** [comment]
ok, thanx for the feedback so far. i'll keep the names as they are, so the old and new can live side-by-side.

in case we want to make these new scrapers the default ones on all installs, someone would need to write some migration code to update guisettings.xml on existing kodi installs.

@DaveTBlake no hurry, plenty of time left to go over all of the details.

@zag2me comments are already in place. see the default.py / scraper.py files

**Comment by MartijnKaijser** [comment]
What i meant was that the id of the old and new scraper should be identical and not start with a new ID.

**Comment by ronie** [comment]
ah ok, got it.

well yes, if we don't rename, we need some migration code.

**Comment by olympia** [comment]
Why no renaming? I actually think we should have different ID and renaming as well. Let's distinguish/ separate old and new and make new default.

Can someone confirm XML scraper engine will enjoy continued support from core?

**Comment by ronie** [comment]
what renaming i mean changing the addon id of the new addons to the ones currently used by the universal scraper.
not sure how to interpret `we should have different ID and renaming as well` ?

**Comment by olympia** [comment]
No, if it was up to me, I would add a different name, e.g. by adding NG (next generation) to the end of the name and a different ID. Treat it as a separate scraper and make it default. Why the new scrapers should use the same id? To update to the new ones having the user believe it's the same scraper? That's not a nice approach IMO.

I don't think we necessarily need an update path, unless XML scrapers will stop working. Or is this the plan? Ditch XML scraper engine completely? Still no feedback on this part of my question yet.

**Comment by ronie** [comment]
nothing has been set in stone, but the point has been raised that keeping xml scraping alive will stand in the way of improving the scraper engine.

File: addons/metadata.integral.artists/lib/musicbrainz.py
Comment (DaveTBlake) [inline addons/metadata.integral.artists/lib/musicbrainz.py]: Can "type", "gender", and "disambiguation" from the Musicbrainz results be set in the artist data too. Also needs adding to wherever artistdata is passed back to CArtist

File: addons/metadata.integral.artists/lib/scraper.py
Comment (DaveTBlake) [inline addons/metadata.integral.artists/lib/scraper.py]: Seems it allready has type and disambiguation properties, but needs "gender" too

File: addons/metadata.integral.artists/lib/scraper.py
Comment (DaveTBlake) [inline addons/metadata.integral.artists/lib/scraper.py]: Is the addon limiting what art types get returned from those in the results returned from the sites? Is that needed?

**Comment by DaveTBlake** [comment]
I think the question of  the future XML scrapers needs answering separately. As best I can tell the scrapers here do not need any changes to the core scraper implementation that will break the XML scrapers, which is just as well because we need to run both for testing purposes. 

If in the future we want to enchance the scraping facilities then it seems the XML scrapers may need to be deprecated or it could hold back what we can do. But we are not there yet, and I hope that @olympia can be encouraged to continue his great work supporting the XML scrapers until we have a good reason to drop them i.e. some really exciting enhanced scraping that needs such core changes we can't continue with them. 

After all the great work @olympia has put in supporting the xml scrapers (when the rest of us couldn't bare to mess with regexp of json) I am disappointed with how recent discussion has gone. Saying almost as a side remark, "oh yes let's deprecate the xml scrapers in v19, but don't worry Olympia you can still support users on previous versions. Why deprecate? Well we _might_ want to do scraping differently".

Meanwhile @ronie I have just made a first pass and bit of testing of the artists scraper, there will be plenty of other things to tweak I'm sure. Lets use this PR to get the Python music scrapers into good shape, not make major strategy commitments.

File: addons/metadata.integral.artists/lib/musicbrainz.py
Comment (DaveTBlake) [inline addons/metadata.integral.artists/lib/musicbrainz.py]: From Musicbrainz documentation https://musicbrainz.org/doc/Artist artist type  can be "Person", "Group", "Orchestra", "Choir", "Character", or "Other" although I would not like to guarentee they stick to that.  Types "Orchestra" and "Choir" need to be treated like "Group" when it comes to having "formed" date rather than "born" etc. 

Also I'm not sure if every artist has a type, and if it doesn't but still has begin or end data then it would still be good to store it somewhere

**Comment by DaveTBlake** [comment]
Artist name is far from unique, and I just hit an issue with fetching artist data from all the sites and merging it - you can end up with a mix of data from different artists! This can also happen when mbid is known, Musicbrainz details will be accurate (so will fanart.tv etc. as it uses mbid) but results from allmusic will be by name only, and that could be a different artist.

Either need some way to check what by artist name requets have all returned is the same artist (comparing other data values perahps?) or at very least a way to avoid fetching data from certain sites when you scrape. For example with the xml artist scraper I turn off allmusic requests because they return the wrong artist too many times.

**Comment by olympia** [comment]
the XML UMS is using allmusic link provided by musicbrainz and it never searches on allmusic itself, so I don't know how it can return the wrong artist. @DaveTBlake, do you have an example?

@ronie, this could be a way for the python scraper. Or does it totally independently searches all the sites? This will be error prone.

**Comment by DaveTBlake** [comment]
>the XML UMS is  using allmusic link provided by musicbrainz 
Interesting, then I think the Python scraper needs to do the same to ensure that results are for the same artist. This applies to lookup by name too, pick the MB entry and use link. 

@olympia I don't have an example of the XML UAS scraper giving the wrong artist data from allmusic. I was possibly mis-remembering my reasons for not picking allmusic as the source for data, it could have been matching by name going wrong at Musicbrainz and hence cascading to all remote sites. Now most of my music has mbids so I don't see the issue. Other reason for avoiding allbase was the muddle of style, mood and genre data it can return. On the XML scrapers you can just choose not to have these fields.

File: addons/metadata.integral.artists/lib/musicbrainz.py
Comment (ronie) [reply addons/metadata.integral.artists/lib/musicbrainz.py]: ok, i've added 'type', 'genre' & 'disambiguation'.

the only other source we use for searching is discogs.com, it does not provide this info

File: addons/metadata.integral.artists/lib/scraper.py
Comment (ronie) [reply addons/metadata.integral.artists/lib/scraper.py]: thx, added.

File: addons/metadata.integral.artists/lib/scraper.py
Comment (ronie) [reply addons/metadata.integral.artists/lib/scraper.py]: please elaborate... i'm not sure i understand what you mean

File: addons/metadata.integral.artists/lib/musicbrainz.py
Comment (ronie) [reply addons/metadata.integral.artists/lib/musicbrainz.py]: spot on. will fix that :-)

yes it's possible an artist does not have a type, but storing metadata 'somewhere' is currently not supported by kodi ;-)

**Comment by ronie** [comment]
very valid point indeed.

i did consider using musicbrainz provided links to allmusic/discogs at one point during the long dev cycle of these scrapers. 
if memory serves my well, i decided not to use them because those links are missing sometimes.
which would mean no allmusic/discogs metadata and artwork would be scraped at all.

also, it's a tiny bit slower to have to wait for musicbrainz to return those results before you can query allmusic/discogs. i'm currently querying all sites simultaneously.

guess that's been the biggest struggle so far... finding the right balance between speed, accuracy and completeness.

but i do agree getting a mix of data from multiple artists is not a good thing either... i'm going to look into it again.

File: addons/metadata.integral.artists/lib/scraper.py
Comment (ronie) [reply addons/metadata.integral.artists/lib/scraper.py]: ... it currently returns only 1 clearart image, instead of all available clearart images (if that's what you mean)

i fixed that a few days ago, but hadn't updated this PR with those changes.

Comment (DaveTBlake) [reply addons/metadata.integral.artists/lib/scraper.py]: Yes returning only 1 image of each type was part of it, I was also unclear that it was saving all the aspects (except for fanart) in `artist.thumbs` property not just _thumb_

Also for art types other than thumb and fanart the art of that type for the artist is being set to the first fetched, overwriting any image that user may have chosen. I'm not sure that the scraper should be setting art, and if it does then only filling the gaps not overwriting chosen art.

File: addons/metadata.integral.artists/lib/musicbrainz.py
Comment (DaveTBlake) [reply addons/metadata.integral.artists/lib/musicbrainz.py]: How about if type = "" or something unexpected then save "begin" and "end" from Musicbrainz to "Born" and "Died" as if it was a person rather than discarding the data.

File: addons/metadata.integral.artists/lib/scraper.py
Comment (DaveTBlake) [inline addons/metadata.integral.artists/lib/scraper.py]: Can all the known properties be set here not just genre, born and relevence, especially "disambiguation" when result is from MusicBrainz. When lookup is by name and many artists found a dialog is displayed to let the user choose which is the artist. Currently this will show genre and born as well as name, but usually ends up as just a list of names (most the same) which is not helpful. I'll need to change core to build better strings from the artist data, but this data also needs to be passed back from the scraper

Comment (ronie) [reply addons/metadata.integral.artists/lib/scraper.py]: thx, slipped my mind to add those properties there as well.

**Comment by ronie** [comment]
scrapers updated with all the suggestions made above

File: addons/metadata.integral.artists/lib/musicbrainz.py
Comment (DaveTBlake) [inline addons/metadata.integral.artists/lib/musicbrainz.py]: Could we save "begin" and "end" from Musicbrainz to "Born" and "Died" as a catch all rather than only when type is Person of Charater, covering when no type or some new type Musicbrainz start using, rather than lose data

File: addons/metadata.integral.artists/lib/scraper.py
Comment (DaveTBlake) [reply addons/metadata.integral.artists/lib/scraper.py]: Checked and this is causing scraping to overwrite previously chosen art. Filling the gaps, as happens for thumbs and fanart, is done by the scraper core code. All the scraper itself has to do is fill the artist.thumbs property, listitem.setArt(art) is not wanted.

**Comment by DaveTBlake** [comment]
Thanks for this @ronie , you can usefully pick up this https://github.com/xbmc/xbmc/pull/15773 during testing which makes more use of the artist data you are passing back.

File: addons/metadata.integral.artists/lib/scraper.py
Comment (ronie) [reply addons/metadata.integral.artists/lib/scraper.py]: @DaveTBlake i've removed listitem.setArt(art) from the scraper and scanned a few albums to the library.
but the result is that lots of art is now missing in the skin (clearart, banner, back, spine, ...).
these art types are also no longer listed in the choose art dialog.

File: addons/metadata.integral.artists/lib/musicbrainz.py
Comment (ronie) [reply addons/metadata.integral.artists/lib/musicbrainz.py]: currently, only the type="other" is missing. 
this type is used for things like a collection of nature recordings (for instance):
https://musicbrainz.org/ws/2/artist/51118c9d-965d-4f9f-89a1-0091837ccf54?inc=url-rels+release-groups&type=album&fmt=json

whenever musicbrainz updates their api (last time was 2017, next update is scheduled for may, 13th) we should review the changes and update the scrapers accordingly (if needed).

File: addons/metadata.integral.artists/lib/scraper.py
Comment (rmrector) [reply addons/metadata.integral.artists/lib/scraper.py]: @ronie do you have the extended artwork types configured in [advancedsettings.xml](https://kodi.wiki/view/Advancedsettings.xml#Music_library_extra_artwork)? That should handle all artwork assigned.

For art types not listed in the choose art dialog, that's because there aren't any assigned in your library yet - possibly because of the above. I've got a change in the works to add all available image types to the list; I should have a PR this week.

**Comment by DaveTBlake** [comment]
Just a note to myself really: when scraping from NFO file CNfoFile loops though all the installed and enabled scrapers to find one that can read the NFO see https://github.com/xbmc/xbmc/blob/eb161cdb7c4c97e9f054192c9ea9e8d00839c305/xbmc/NfoFile.cpp#L78

Since the Python scrapers can do this, the intention is that it is done by local information scraper which can, but from user experiences it seems that sometimes the one of the other xml scrapers can be called instead hence seeing xml scrapers used in the log or raising parsing errors because of unexpected xml in the nfo file.

I'm not completely clear on this yet, but there is some issue with scraping NFO files and having Python scrapers enabled

File: addons/metadata.integral.albums/lib/scraper.py
Comment (enen92) [inline addons/metadata.integral.albums/lib/scraper.py]: please drop `thumbnailImage` and use `setArt`

File: addons/metadata.integral.artists/lib/scraper.py
Comment (enen92) [inline addons/metadata.integral.artists/lib/scraper.py]: same here

**Comment by ronie** [comment]
thx, updated.

**Comment by DaveTBlake** [comment]
A bit of incompatibility with NFO scraping @ronie, can you stop both scrapers from returning available art urls that have  `spoof="" cache="" `. E.g.
```
<thumb spoof="" cache="" aspect="thumb">http://coverartarchive.org/release/59b5a40b-e2fd-3f18-a218-e8c9aae12ab5/5270052042.jpg</thumb>
```
I can't see where it comes from but if results with that additional syntax are saved to an NFO file then 
it prevents the NfoURL xml scraper from running correctly. 

The example results in a request to Musicbrainz
`https://musicbrainz.org/ws/2/release/59b5a40b-e2fd-3f18-a218-e8c9aae12ab5/5270052042.jpg`
to resolve mbid that obviously fails.

**Comment by ronie** [comment]
@DaveTBlake i'm not setting those attributes. spoof/cache is added by core:
https://github.com/xbmc/xbmc/blob/73b6095e7e3691fd6ca988144e68c0f6886b4097/xbmc/utils/ScraperUrl.cpp#L328-L332

**Comment by DaveTBlake** [comment]
Thanks for finding where it was coming from @ronie, I'll check it out. Must be something the xml scraper doesn't use, at least it doesn't end up in the results

**Comment by phunkyfish** [comment]
@ronie will you be progressing this further, closing it or put it on the backburner?

**Comment by ronie** [comment]
yes, it's still a candidate for v19 inclusion.

**Comment by DaveTBlake** [comment]
@ronie prod me to give this a good testing (again) soon, too busy with life right now. I am also hoping to do the art type whitelist etc. changes to go with it. This PR is open on my browser to remind me, but...

**Comment by ronie** [comment]
thanx @DaveTBlake me's patiently waiting for you to do the art whitelist as well ;-)
glad to hear it hasn't slipped your mind.

**Comment by ronie** [comment]
as suggested during the last team meeting, set the python based scrapers as default on new kodi installs.

**Comment by jjd-uk** [comment]
Perhaps time to @DaveTBlake  another prod then

**Comment by DaveTBlake** [comment]
I finally found some time to do some detailed testing and a careful examination of the album scraping results (yet to do same for artists). Although well overdue it was worthwhile as it did show up a minor (I hope) flaw. 

When searching on artist and album title (no mbids) it sometimes picks singles or EPs from the Musicbrainz db rather than a release group and release for an album. Examples for testing:
"Standing Stone", "Paul McCartney"
https://musicbrainz.org/taglookup?tag-lookup.artist=Paul+McCartney&tag-lookup.release=Standing+Stone&tag-lookup.tracknum=&tag-lookup.track=&tag-lookup.duration=&tag-lookup.filename=
The results return the release as a single first in the list.

Other examples
"Love Scenes",	"Beverley Craven"
"Rise", "Gabrielle"
"If You See Him", "Reba McEntire"

I don't know if you can limit the lookup to the Musicbrianz  primary "releasetype" of album (and still include compliations and other secondary types) or if it is matter of checking the results when there is more than one to return the most likely i.e. album not single or EP

**Comment by ronie** [comment]
> Examples for testing:
> "Standing Stone", "Paul McCartney"

the scraper will use this query:
https://musicbrainz.org/ws/2/release/?query=release:%22Standing%20Stone%22%20AND%20(artistname:%22Paul%20McCartney%22%20OR%20artist:%22Paul%20McCartney%22)&fmt=json

it uses the match 'score' from musicbrainz to return the best match.
the single has a score of '100' and the album only '71'.
i don't know how they are scoring results, but according to musicbrainz the album artist is _Paul McCartney; London Symphony Orchestra, Lawrence Foster_ so that might be a reason for the low score...

> "Rise", "Gabrielle"

https://musicbrainz.org/ws/2/release/?query=release:%22Rise%22%20AND%20(artistname:%22Gabrielle%22%20OR%20artist:%22Gabrielle%22)&fmt=json

in this case both the single and the album have a match score of '100'.
for a situation like that, the scraper checks which release-group contains to most releases and pick that one. which happens to be the single.
i agree i'd be a good idea to throw an 'album vs 'single' evaluation into the mix as well if there a tie.
and perhaps even the release status (official vs. bootleg) as well...

though it might be tricky/complicated to apply our own scoring system in top of the musicbrainz one.
what if a bootleg/single has a mb score of '100' and the official/album has a mb score of '99'


> "Love Scenes", "Beverley Craven"

both single and album score 100'. both release groups have an equal number of releases.
so currently the scraper will simply return the first result from musicbrainz (which happens to be the single)

> "If You See Him", "Reba McEntire"

same as above (EP vs album)

> Oh and a tech reminder (just in case) Musicbrianz release type is put into album.type. There is an album.releasetype too, but that is an unfortunately named internal flag and nothing outside core processing should set it.

thanx, i happened to come across that a few days ago as well, and have already made some fixes (not pushed yet).

**Comment by DaveTBlake** [comment]
Thanks @ronie, no idea how MB do their scoring either, must have been luck that the XML scraper picked the right ones.  Having the release type returned did make it easier to spot. 

Artist name (especially if only one of many album artists given) and album just isn't always enough to uniquely identify even the release group. For example Peter Gabriel has released 3 totally different albums with different songs all called "Peter Gabriel".  Great idea when scores are the same to use the release-group that contains the most releases, it is a reasonable best guess.

If being offered results that include multiple release groups with exactly matching album  but various release types then I think it would be good if we prefer those with primary type "album" over the others regardless of how many releases there are.  Generally albums in the music library are album releases, so reasonable to assume that as default when we don't have mbids. 

But say a user has added an EP, multi-track "single"  or broadcast to their music library as an album and there is an album release with the same title, then it is up to the user to tag it with Musicbrainz IDs otherwise the album scraper will match it with the album release.

For that reason too I would not worry about release status. If all we have is artist and album name then we return the most likely (score and release count) _album_ release.

>>There is an album.releasetype too, but that is an unfortunately named internal flag and nothing outside core processing should set it.

>thanx, i happened to come across that a few days ago as well, and have already made some fixes (not pushed yet).

Yes I noticed that `DetailsFromFileItem` in Scraper.cpp sets album.releasetype and it shouldn't. Were you fixing that, or did you just mean in the addon so shall I raise a PR for core?

**Comment by DaveTBlake** [comment]
Now done a detailed testing and a careful examination of the artist scraping results - looking good.

Only question I have is about the return of more than just year (but month and day sometimes too) in the disography entries. I need to look at bit closer at what core does with that data. The discography is not the best artist feature, and there could easily be improvements to be made core side. Could the scraper easily fetch more disco data, in particluar release group mbid? It could make matching the disco items to the albums already in the library easier and thus a better user experience.

**Comment by ronie** [comment]
if returning to full date for the artist discography is an issue, i could ofc limit it to only returning the year.

returning the release-group mbid would not be a problem, i have that info available.

here's an example of the available data for the artist discography from musicbrainz:
disco

**Comment by DaveTBlake** [comment]
>if returning to full date for the artist discography is an issue, i could ofc limit it to only returning the year.

Not a problem, held as string

>returning the release-group mbid would not be a problem, i have that info available.

That would be useful. It needs a core change to make use of it, but certainly pass it back for each discography item

**Comment by DaveTBlake** [comment]
Artists scraper: 

I have made the core changes to make use of the artist scraper returning release-group mbid for the artist discography entries. Stacked on top my recent PR and your album scraping ones in a branch on my repo https://github.com/DaveTBlake/xbmc/tree/DiscoWithMbid if you are interested.

I think it works, but need a version of the artists scraper that sets ".musicbrainzreleasegroupid" as well as ".title" and ".year" to test and be sure.

**Comment by ronie** [comment]
ok, i've added musicbrainzreleasegroupid to the artist scraper (v1.0.6)
it's in my upstream repo: https://gitlab.com/ronie/metadata.generic.artists

i only briefly tested it. the musicbrainzreleasegroupid are indeed added to the discography table.
one bug though, if an artist has released multiple different albums with the same name, now only 1 will show up in the musicinfo dialog.
tested it with this artist: https://musicbrainz.org/artist/6fe07aa5-fec0-4eca-a456-f29bff451b04

**Comment by DaveTBlake** [comment]
Thanks for adding musicbrainzreleasegroupid

>if an artist has released multiple different albums with the same name, now only 1 will show up in the musicinfo dialog.

**Comment by DaveTBlake** [comment]
@ronie I have done some more scraper tests on fetching artists by name (without Musicbrainz id), here are a few oddities you might check out. 

"Paul McCartney"  does not scrape, weird because he is definitely in the Musicbrainz database, website search on artist name works as does XML query
http://musicbrainz.org/ws/2/artist/?query=artist:Paul%20McCartney
Giving link https://musicbrainz.org/artist/ba550d0e-adac-4864-b88b-407cab5e76af

More artists that were party scraped but didn't get Musicbrainz data despite being in their database:
"James McCarty" he is in Musicbrainz https://musicbrainz.org/artist/ba900218-68b3-434d-8552-989d8441bba3  The probelm could be that is an alias name and the main entry is "Jim McCarty", can you search alias names?

Maybe similarly alias examples
"Northern Sinfonia of England" again an alias of Royal Northern Sinfonia
https://musicbrainz.org/artist/b4cd94be-c8d8-48ab-8bf5-4eba5fb69f98/aliases
"National Symphony Orchestra Of Ireland" alias RTÉ National Symphony Orchestra
https://musicbrainz.org/artist/1348b9ee-16de-4d09-8734-9c4b930e8458/aliases

"The Choir Of York Minster" https://musicbrainz.org/artist/a389e08e-2df0-4ade-8de7-66af0dc3b213
Website based search seems to ignore the "the" and finds the artist

There may well be nothing you can do about matching on alias names, and TBH info on orchestras is pretty sparse and futile anyway, but I thought I would mention it in case there was an API option or the script was getting the results and then skipping at as request and result didn't match (check any alias names too?).

Finally a couple of artists where the record got marked as scraped but no info was fetched:
"Desmond Wright", "Anne Chaurand"
Both just had `<fanart />` added to the strFanart field of the album table
Neither have Musicbrainz entries, and they are obscure so I would not expect them to scrape anywhere else, but they do have discogs entries 
https://www.discogs.com/artist/5666198-Anne-Chaurand
https://www.discogs.com/artist/4300294-Desmond-Wright

Can you handle where the scraper does find an enrty but none of the details we want are returned as being same as not scraped? We didn't get any info, and once marked as scraped attempts to batch rescape exclude it.

**Comment by ronie** [comment]
hey @DaveTBlake, when it comes to artist scraping, i'm going to need a debuglog.

even though you have not tagged your files with an artist mbid, the album scraper will return it.
kodi will then pass this mbid to the artist scraper.
so if an album is found on musicbrainz, kodi does have the artist mbid.

>Finally a couple of artists where the record got marked as scraped but no info was fetched:
"Desmond Wright", "Anne Chaurand"

for "Desmond Wright" the artist discography should be scraped. could you check?
for "Anne Chaurand", discogs doesn't hold any additional data. there's one album listed, but she's not credited as the main artist, so it will not be included in the artist discography.


> Can you handle where the scraper does find an enrty but none of the details we want are returned as being same as not scraped?

not really, one thing the artist scraper will always scrape and return is the 'artist name'.

**Comment by DaveTBlake** [comment]
Ha, found the "Paul McCartney" issue on the music db side, so something for me to work on :) 

>for "Desmond Wright" the artist discography should be scraped. could you check?

Yes, have a discography for "Desmond Wright"

Considering it there has always been the possibility of scraper identifying an artist (or album) but not having any additional info to return,  "Anne Chaurand" is the first example I have seen. Should such artists/albums be marked as scraped? As a user I would like to be able to retry at a later date as a batch operation for those items where I didn't get any info regardless of  why it didn't get it (the wider community could have added some data). If the scraper can't catch that situation easily then it is something that could be done in core, but it doesn't happen that often, so maybe such checks aren't worth it?

>even though you have not tagged your files with an artist mbid, the album scraper will return it.
kodi will then pass this mbid to the artist scraper. So if an album is found on musicbrainz, kodi does have the artist mbid.

Yes I know, but in that test I only ran the artist scraper from "query info for all" on an artists node. Separate scraping giving the scraper most chance to mis-identify things. 

The artist mbid was not returned for "James McCarty",  "Northern Sinfonia of England", "National Symphony Orchestra Of Ireland" and "The Choir Of York Minster" despite there being entries in Musicbrainz. That and lack of other data makes me think that the scraper didn't find the Musicbrainz entry. Maybe it just can't, but worth checking. I'll repeat the scrape and get some debug logs for you.

**Comment by DaveTBlake** [comment]
In getting logs I decided to rescan with "fetch additional info on update" enabled, hence using both scrapers in combination. It kicked out a few new Python errors @ronie for you to check out. Rather large log here 
https://drive.google.com/file/d/14uOhArar1AtfRhlozm18qJoGp13tdbZ4/view?usp=sharing
Using build from my repo that has your album scraping PRs and my recent PR + enhanced discography stuff hence MyMusic79 (true bleeding edge).   
There are some contraint db errors too that I'm dealing with separately. 

There are 13 albums which it failed to scrape this run that it had scraped OK before from when run "query info for all" with an earlier version of the addon (I haven't checked the log for each yet). Not all albums are scrapeable because they of how they are tagged, so some failures are expected.

It is worth noting that "James McCarty" has picked up mbid from the combined album and artist scrape, which shows how useful doing that can be.

**Comment by DaveTBlake** [comment]
@ronie a smaller log with the v1.0.6 album scraper errors https://paste.kodi.tv/ibesuvobiy.kodi

**Comment by ronie** [comment]
> Considering it there has always been the possibility of scraper identifying an artist (or album) but not having any additional info to return, "Anne Chaurand" is the first example I have seen. Should such artists/albums be marked as scraped? As a user I would like to be able to retry at a later date as a batch operation for those items where I didn't get any info regardless of why it didn't get it (the wider community could have added some data). If the scraper can't catch that situation easily then it is something that could be done in core, but it doesn't happen that often, so maybe such checks aren't worth it?

hmm, when is the data returned by the scraper complete enough to be considered a successful scrape?
for instance, let's say the only piece of info it can find is 'artist.gender'.

> https://drive.google.com/file/d/14uOhArar1AtfRhlozm18qJoGp13tdbZ4/view?usp=sharing

wrong link ;-)

> @ronie a smaller log with the v1.0.6 album scraper errors https://paste.kodi.tv/ibesuvobiy.kodi

thanx, i'll fix those.

**Comment by DaveTBlake** [comment]
>hmm, when is the data returned by the scraper complete enough to be considered a successful scrape? for instance, let's say the only piece of info it can find is 'artist.gender'.

It was being "scraped" but returning no data at all that surprized me. Mostly I just need to update my mental landscape to include that possibility, the rare "scraped but fetched nothing" is the least of user issues. : )

**Comment by ronie** [comment]
> can you search alias names?

yes, that is something i can do. i'm unsure though if that would increase the likelihood of false-positives or not.  

it would however fix your issue with these two:
- James McCarty
- National Symphony Orchestra Of Ireland



> "Northern Sinfonia of England" again an alias of Royal Northern Sinfonia
> https://musicbrainz.org/artist/b4cd94be-c8d8-48ab-8bf5-4eba5fb69f98/aliases

it does not list "Northern Sinfonia of England" as an alias?
searching the api also yields no results:
https://musicbrainz.org/ws/2/artist/?query=alias:%22Northern%20Sinfonia%20of%20England%22&fmt=json



> "The Choir Of York Minster" https://musicbrainz.org/artist/a389e08e-2df0-4ade-8de7-66af0dc3b213
> Website based search seems to ignore the "the" and finds the artist

yup, but the api doesn't
https://musicbrainz.org/ws/2/artist/?query=artist:%22The%20Choir%20Of%20York%20Minster%22&fmt=json
in this case there's nothing i can do.

**Comment by DaveTBlake** [comment]
Artist scraper.

>>    can you search alias names?

>yes, that is something i can do. i'm unsure though if that would increase the likelihood of false-positives or not.

As for more risk of false-positives I don't know, maybe the MB score algorithm handles this well?

Musicbrainz xml web service docs say 
"Artist search terms with no fields specified search the artist, sortname and alias fields. "
see https://musicbrainz.org/doc/Development/XML_Web_Service/Version_2/Search
and the xml scraper also finds "James McCarty", so I think that it has always done an alias name search.  Does that substantiate that is isn't too much a  false-positives risk?

Perhaps try it and see?

Thanks for checking through my examples. It is not that I expect everything to scrape and perfect results, I just raised them in case these fringe case could teach us something.

**Comment by ronie** [comment]
> Perhaps try it and see?

yup, it's what i did. i have tested it the last 2 days and did not find any regressions.

both scrapers have been updated to v1.0.7
in the album scraper, the crashes from your previous log have been fixed
the artist scraper will now search for name / alias / sortname

**Comment by ronie** [comment]
ready to merge.

**Comment by DaveTBlake** [comment]
@ronie I hope to do one more big test run over this weekend combined with the other scraper improvements that are now in master etc. Maybe hold for that before merge?

**Comment by ronie** [comment]
sure, not a problem.
my aim is to get it included in the alpha1 release.

**Review by DaveTBlake — APPROVED** [review]
Finally got that testing done @ronie rolled it into testing the "better art out the box" stuff I am working on. All looking good  :)

**Comment by ronie** [comment]
great, thanx a million for all the help and testing so far!

if you need someone to test-drive your "better art out the box" work, you know where to find me :-)
(and it likely requires updates to these scrapers)

**Pull request merged**
